### PR TITLE
CB-6680: repair after upgrade

### PR DESCRIPTION
**Pull request by pdarvasi**
This PR contains the following improvements:
- CB-6680 - Replace VM image after a successful in-place upgrade
- CB-6795 - Unify OS upgrade flow and CDP upgrade flow

Besides those, the commit is meant to fix the following issues
- CB-7841 - Parameter --lock-components should work in conjunction with --imageid
- CB-7842 - OS upgrade should not be permitted to image with former timestamp

### CODE REVIEW

**Comment by pdarvasi** [comment]
retest this please

File: core-api/src/main/java/com/sequenceiq/cloudbreak/validation/UpgradeRequestValidator.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        }
+        return isOsUpgrade(value) || isRuntimeUpgrade(value);
+    }
+
+    private boolean isOsUpgrade(UpgradeV4Request request) {
+        return Boolean.TRUE.equals(request.getLockComponents()) && StringUtils.isEmpty(request.getRuntime());
+    }
+
+    private boolean isRuntimeUpgrade(UpgradeV4Request request) {
+        return !Boolean.TRUE.equals(request.getLockComponents())
```
Comment (keyki) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/validation/UpgradeRequestValidator.java:27]: !true we can use false?

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java [hunk 2]
```diff
@@ -72,6 +74,17 @@ private ImageFilterResult filterImages(List<Image> availableImages, Image curren
         return new ImageFilterResult(new Images(null, images, null), getReason(images));
     }
 
+    private Predicate<Image> filterPreviousImages(Image currentImage, boolean lockComponents) {
+        return image -> {
+            boolean result = !lockComponents
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java:79]: Can you add some parentheses,  what are we comparing?

File: datalake-api/src/main/java/com/sequenceiq/sdx/validation/UpgradeRequestValidator.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        }
+        return isOsUpgrade(value) || isRuntimeUpgrade(value);
+    }
+
+    private boolean isOsUpgrade(SdxUpgradeRequest request) {
+        return Boolean.TRUE.equals(request.getLockComponents()) && StringUtils.isEmpty(request.getRuntime());
+    }
+
+    private boolean isRuntimeUpgrade(SdxUpgradeRequest request) {
+        return !Boolean.TRUE.equals(request.getLockComponents())
```
Comment (keyki) [inline datalake-api/src/main/java/com/sequenceiq/sdx/validation/UpgradeRequestValidator.java:27]: !true = false

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxUpgradeService.java [hunk 4]
```diff
@@ -290,8 +241,4 @@ private boolean stackAndClusterAvailable(StackV4Response stackV4Response, Cluste
                 && cluster.getStatus() != null
                 && cluster.getStatus().isAvailable();
     }
-
-    private String getMessage(String imageId) {
-        return messagesService.getMessage(ResourceEvent.DATALAKE_UPGRADE.getMessage(), Collections.singletonList(imageId));
```
Comment (keyki) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxUpgradeService.java:295]: messagesService is not used anymore

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java [hunk 5]
```diff
@@ -24,18 +27,23 @@
...
 import com.sequenceiq.datalake.service.sdx.SdxService;
 import com.sequenceiq.flow.api.model.FlowIdentifier;
+import com.sequenceiq.sdx.api.model.SdxUpgradeReplaceVms;
 import com.sequenceiq.sdx.api.model.SdxUpgradeRequest;
 import com.sequenceiq.sdx.api.model.SdxUpgradeResponse;
 
 @Component
 public class SdxRuntimeUpgradeService {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(SdxUpgradeController.class);
```
Comment (tiborpopovics) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java:45]: Please rename the logger class to `SdxRuntimeUpgradeService `

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java [hunk 6]
```diff
@@ -56,51 +64,63 @@
...
     }
 
-    public SdxUpgradeResponse checkForRuntimeUpgradeByCrn(String userCrn, String crn, SdxUpgradeRequest upgradeSdxClusterRequest) {
-        verifyRuntimeUpgradeEntitlement(userCrn);
-        SdxCluster sdxCluster = sdxService.getByCrn(userCrn, crn);
-        UpgradeV4Response upgradeV4Response = stackV4Endpoint.checkForClusterUpgradeByName(WORKSPACE_ID, sdxCluster.getClusterName(),
-                sdxUpgradeClusterConverter.sdxUpgradeRequestToUpgradeV4Request(upgradeSdxClusterRequest));
-        return sdxUpgradeClusterConverter.upgradeResponseToSdxUpgradeResponse(upgradeV4Response);
+    public SdxUpgradeResponse checkForUpgradeByCrn(String userCrn, String crn, SdxUpgradeRequest upgradeSdxClusterRequest) {
+
```
Comment (tiborpopovics) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java:74]: Please remove this empty line

File: core-api/src/main/java/com/sequenceiq/cloudbreak/validation/UpgradeRequestValidator.java (hunk 1 above)
Comment (pdarvasi) [reply core-api/src/main/java/com/sequenceiq/cloudbreak/validation/UpgradeRequestValidator.java:27]: This handles "null" value as Runtime upgrade

File: datalake-api/src/main/java/com/sequenceiq/sdx/validation/UpgradeRequestValidator.java (hunk 3 above)
Comment (pdarvasi) [reply datalake-api/src/main/java/com/sequenceiq/sdx/validation/UpgradeRequestValidator.java:27]: This handles "null" value as Runtime upgrade

**Pull request merged**